Make native_batch_norm structured

wconstab opened the pull request.

bdhirsh commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -19,6 +19,34 @@ static const int MIOPEN_DIM_MAX = 5;
+  set_output(0, input.sizes(), input.options().memory_format(input.suggest_memory_format()));
+
+  auto out_stats_options = input.options();
+  if (input.scalar_type() == c10::ScalarType::BFloat16 || input.scalar_type() == c10::ScalarType::Half) {
```
ok, so I think this logic comes from the original call to `accumulateType<...>` in the code, right? https://github.com/pytorch/pytorch/pull/67343/files#diff-e895801158bc7155ba3b61826cdd129c5e2caa5c3519171abfbe9bbe5406f52fL435.

We probably want to re-use that logic: if I'm following it correctly, it looks like the accumulate type really is the output dtype of the stats (because those stats are globals that are re-used throughout multiple calls to batch_norm).

And... I don't think there's anything backend specific about `Half/BFloat16`. I just audited it, and it looks like the [only difference happens for `float`](https://github.com/pytorch/pytorch/blob/603116a6aef8027fb363e7f081702e00cf603ff2/aten/src/ATen/AccumulateType.h#L60) (float + cpu -> float, float + cuda -> double).

So... I think maybe just re-use that logic, and add a comment calling out that we have to pass `is_cuda` into the template because cuda can have different output dtypes (and specifically mention the float->double special case for cuda for the stats). Does that sounds reasonable?

bdhirsh commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -19,6 +19,35 @@ static const int MIOPEN_DIM_MAX = 5;
+  if (input.device().type() == c10::kCUDA){
+    // This meta func matches the non-uniform dtype promotion done in CPU/CUDA impls.
+    // It uses a larger type for stats accmulation on CUDA (half/bf16->float, float->double)
+    out_stats_options = out_stats_options.dtype(
```
This sounds ok to me. Maybe just add something a bit more explicit, like:
```
Meta functions expect all dtype + shape calculations to be backend independent. However, the dtype inference logic for batch_norm is NOT backend-independent: on CUDA, we use a larger type for stats accumulation (half/bf16->float, float->double).
This means that if you are using meta tensors for shape/dtype inference, then your dtype outputs might not match the actual outputs when run on CUDA.
```

wconstab commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -627,32 +668,45 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cpu(
-  const Tensor& running_mean = c10::value_or_else(running_mean_opt, [] {return Tensor();});
-  const Tensor& running_var = c10::value_or_else(running_var_opt, [] {return Tensor();});
-
-  checkBackend("batch_norm_cpu", {self, weight, bias, running_mean, running_var}, Backend::CPU);
```
@bdhirsh I am suspicious that omitting this check is the reason for the last CI failure i'm still facing.  

https://www.internalfb.com/intern/opensource/ci/job/log/3080152672308172/
2021-11-09T00:36:51.5070547Z AssertionError: "Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU'" does not match "Could not run 'aten::empty.memory_format' with arguments from the 'QuantizedCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::empty.memory_format' is only available for these backends: [CPU, CUDA, Meta, MkldnnCPU, SparseCPU, SparseCUDA, SparseCsrCPU, SparseCsrCUDA, BackendSelect, Python, Named, Conjugate, Negative, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradLazy, AutogradXPU, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode, Functionalize].


It looks like the test (test_fusion.py:90) is expecting a particular failure of native_batch_norm, but that's no longer happening and empty is failing instead.

        with self.assertRaisesRegex(RuntimeError, "Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU'"):
            checkQuantized(model)

I think you had suggested I shouldn't use checkBackend in a device-specific way in the meta-kernel before. But is it OK to have it check specifically for unallowed devices such as quantized?  Or what would be the best way to handle this?

wconstab replied:
ok- do you recommend modifying the unit test then?  or should I try to raise the same error as previously raised from inside meta::native_batch_norm?

bdhirsh replied:
That means that the error message for quantized ops will get a little worse though - we could probably make it better, but that would require some special handling in the codegen.

bdhirsh replied:
Yeah, let me take a look and see how hard it would be to add the check in the codegen. I think it would be reasonable to add it if isn't too much of a hack.

bdhirsh replied:
https://github.com/pytorch/pytorch/pull/71928 should improve the error message (I tried rebasing your changes on top of it and making the error message no longer mentions `empty.memory_format`).

You'll still need to tweak the test a bit because I couldn't easily use the exact same error message though.

bdhirsh replied:
https://www.internalfb.com/diff/D33823417 just landed internally, so you should be able to rebase and (fingers crossed) the error will go away

wconstab replied:
yea i have already rebased the internal diff and verified it works.  (Unfortunately, I'm still stuck on a separate internal error)

bdhirsh commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -19,6 +19,60 @@ static const int MIOPEN_DIM_MAX = 5;
+      && running_mean->is_contiguous()
+      && running_var->is_contiguous();
+
+  // TODO(whc) using input.suggest_memory_format() passes test_ops.py,  fails test_nn.py (:9252, self.assertTrue(out.is_contiguous(memory_format=torch.channels_last)))
```
Good catch!

It looks like that special handling isn't there in the existing cuda kernel though... https://github.com/pytorch/pytorch/blob/f9422e1c6b5dab15eacbbb0e1d4370b07db1e8e8/aten/src/ATen/native/cuda/Normalization.cu#L432. Maybe that means that the bug in the above issue was "fixed" only for cpu?

wconstab replied:
@ngimel Could you weigh in? 
It looks like on cuda, we are just using the input memory format.  On CPU, we are checking _all_ the input tensors, and then making a decision.  Is this safe to land?  It seems like more work should be done to unify the cpu/cuda paths, but i'm hoping to not have to do it all in this PR.

ngimel replied:
So this PR is changing cuda behavior st output will no longer propagate channels last if some of the other tensors are not contiguous? That's not great but we can live with this as long as we can fix it relatively soon (just by calling `.contiguous` on all those tensors in the cpu implementation).

wconstab replied:
i think you're right, that would be a change in cuda behavior with this PR;  

Frankly, I'm not sure what the reasoning behind the current behavior on CPU side is.  Is it better to just make a simpler check (like the one cuda uses) and then change the unit tests?  or are we stuck mirroring whatever the current behavior is, to avoid it being a BC issue?

ezyang commented:
bumping

bdhirsh committed:
improve quantized error checking for structured kernels (#71928)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71928

This PR improves the error messaging for structured operators that don't have a quantized backend implementation.

`at::native_batch_norm` does not currently have a quantized implementation. The current error when you try to run that op with quantized tensors is:
```
"Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU' backend"
```

Making the operator structured (in https://github.com/pytorch/pytorch/pull/67343) makes the error message worse though:
```
Could not run 'aten::empty.memory_format' with arguments from the 'QuantizedCPU' backend
```

Why does that happen? After making `native_batch_norm` structured, it gets a code-generated `CompositeExplicitAutograd` kernel. The codegen'd kernel just constructs an empty tensor using `at::empty()` and calls into the `out=` kernel.

This is useful for most backends because they can choose to implement only the `out=` operator if they want to, and get a composite out-of-place kernel that would work automatically.

This ends up being bad for quantization though because quantization doesn't support `at::empty()`, so we end up breaking with a (pretty useless) error message.

In this PR, I add a dynamic check for if the operator we're trying to call is meant to be quantized, and if it is I return an undefined tensor. This lets us try to dispatch all the way to `native_batch_norm.out`, at which point the dispatcher complains with a (friendlier) error message.

**Testing**

I tested the PR by rebasing https://github.com/pytorch/pytorch/pull/67343 on top of it and running `python test/test_quantization.py TestFuseEager.test_fuse_module_train` to confirm that the error message now mentions `native_batch_norm` instead of `empty.memory_format`

Differential Revision: D33823417

Test Plan: Imported from OSS

Pulled By: bdhirsh

fbshipit-source-id: 6a1b815e9bf6357336524d2934fd3fc4c9de62f4

wconstab committed:
Make native_batch_norm structured (#67343)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/67343

Differential Revision: D32172318

Pulled By: wconstab

fbshipit-source-id: b54014d4cc9ac7873f38cb24a47678f81f34596a
